common/src/test/java/io/netty/util/HashedWheelTimerTest.java

```diff
@@ -149,9 +146,57 @@ public void run(final Timeout timeout) throws Exception {
         for (int i = 0; i < scheduledTasks; i++) {
             long delay = queue.take();
             assertTrue("Timeout + " + scheduledTasks + " delay " + delay + " must be " + timeout + " < " + maxTimeout,
-                    delay >= timeout && delay < maxTimeout);
+                delay >= timeout && delay < maxTimeout);
+        }
+
+        timer.stop();
+    }
+
+    @Test
+    public void testRejectedExecutionExceptionWhenTooManyTimeoutsAreAddedBackToBack() {
+        HashedWheelTimer timer = new HashedWheelTimer(Executors.defaultThreadFactory(), 100,
+            TimeUnit.MILLISECONDS, 32, true, 2);
+        timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        try {
+            timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+            fail("Timer allowed adding 3 timeouts when maxPendingTimeouts was 2");
+        } catch (RejectedExecutionException e) {
+            // Expected
+        } finally {
+            timer.stop();
         }
+    }
 
+    @Test
+    public void testNewTimeoutShouldStopThrowingRejectedExecutionExceptionWhenExistingTimeoutIsCancelled()
+        throws InterruptedException {
+        final HashedWheelTimer timer = new HashedWheelTimer(Executors.defaultThreadFactory(), 100,
+            TimeUnit.MILLISECONDS, 32, true, 2);
+        timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        Timeout timeoutToCancel = timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        timeoutToCancel.cancel();
+        Thread.sleep(200); // Wait for next tick so that cancelled timeouts are processed.
+        timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testNewTimeoutShouldStopThrowingRejectedExecutionExceptionWhenExistingTimeoutIsExecuted()
+        throws InterruptedException {
+        final HashedWheelTimer timer = new HashedWheelTimer(Executors.defaultThreadFactory(), 25,
+            TimeUnit.MILLISECONDS, 4, true, 2);
+        timer.newTimeout(createNoOpTimerTask(), 1, TimeUnit.SECONDS);
+        timer.newTimeout(createNoOpTimerTask(), 90, TimeUnit.MILLISECONDS);
+        Thread.sleep(200); // Wait for timeout to be executed.
```

Comment:
Can we avoid using sleep all together? Perhaps executing a task on the eventloop which completes a future that the test thread blocks on with some timeout?